### PR TITLE
Added R/SQL DBI varchar(max) error solution

### PR DESCRIPTION
**Pull request by rmbielby**
## Overview of changes

Created a new using R to connect to SQL section and moved the existing running SQL queries in R section into it. Added a short intro section on dplyr and then added the error and associated fix for the DBI varchar(max) bug.

## Why are these changes being made?

The SQL in R section was a bit light on some useful topics now that more people are using R in this way. The DBI bug is quite a common one so needed referencing, but also didn't look like we'd really given much attention to dbplyr in the existing guidance.

## Checklist before requesting a review
- [ ] I have checked the contributing guidelines
- [ ] I have checked for and linked any relevant issues that this may resolve
- [ ] I have checked that these changes build locally
- [ ] I understand that if merged into main, these changes will be publicly available

### CODE REVIEW

File: learning-development/r.qmd [hunk 1]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+```
+
+Once you've established a connection, you can then create piped queries as you would with dplyr:
+
+```
+query_result <- collect(
+  tbl(con, Id(schema='dbo', table='table_name')) %>%
+  select(example_id_column, example_filter_column, example_indicator_column) %>%
+  filter(example_filter_column=='Filter entry 1', example_indicator_column <= 120)
+)
```
Comment (cjrace) [inline learning-development/r.qmd:353]: I've always written this with collect separately like the `x <- tbl(...) %>% select(...) %>% collect()`, so you do all the things and then collect the results at the end. I'm happy either way, just curious if it makes any difference or if there's a preferred way we could be consistent with?

File: learning-development/r.qmd [hunk 2]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+```
+
+The key differences to just using dplyr are:
+
+- we use tbl() to point to the SQL table on the SQL database in place of supplying a data-frame
+- we use the collect() function to tell R to collate the commands and run them on the SQL database
+
+### Common issues interacting with a SQL database from within R
```
Comment (cjrace) [inline learning-development/r.qmd:361]: Can we add line breaks around this header for style consistency?

File: learning-development/r.qmd [hunk 3]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+
+```
+Error: nanodbc/nanodbc.cpp:3166: 07009: [Microsoft][SQL Server Native Client 11.0]Invalid Descriptor Index 
+Warning message:
+In dbClearResult(rs) : Result already cleared
+```
+
+There are a couple of solutions to this:
+- avoid creating tables containing varchar(max) columns (i.e. always set a field length for your character fields)
+- make sure all varchar(max) columns are placed last in ay select query that you write
```
Comment (cjrace) [inline learning-development/r.qmd:375]: Think you might need to add a blank line somewhere here (or at least that's my usual fudge - I should probably work out the actual reason / formatting to make it work!), as the bullet points aren't formatting correctly when you build it:
image

**Review by cjrace — CHANGES_REQUESTED** [review]

File: learning-development/r.qmd [hunk 4]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
 
 More tips for moving from using SQL to using R can be found in the [SQL-to-R wiki](https://github.com/dfe-analytical-services/excel-to-R/wiki/SQL-to-R-how-do-I...).  
 
+---
+
+## Interacting with a SQL database from within R-scripts
 
 ---
 
+R can be used to execute SQL scripts to extract data from a database as well as querying the database directly via R. There are two primary ways to do this: 1. executing a SQL script as is within R and 2. using dbplyr to create SQL code for you. Which you use will depend on how comfortable with SQL and R and also if you already have existing SQL scripts that you want to execute or you're writing new database queries.
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:257]: Done

File: learning-development/r.qmd [hunk 5]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+
+### Executing SQL scripts from within R
+
+This is primarily useful if you already have existing SQl scripts that you want to build into an R controlled pipeline. For using R to execute a SQL script you'll need the SQL script/s to be in your R Project and to make a connection via R to the database.
+
+```
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:263]: Done

File: learning-development/r.qmd [hunk 6]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+# Execute SQL query and pull into R ====
+
+my_table <- dbGetQuery(con, getSQL("my_query_script.sql"))
+
+```
+
+If you're struggling to get your SQL scripts to execute from R, try adding the following two lines to the top of your SQL script to force the formatting required for it to work:
+
+
+```
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:315]: Done

File: learning-development/r.qmd [hunk 7]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+
+---
+
+### Using dbplyr to create database queries
+
+If you're familiar with dplyr then using the dbplyr package to interact with a SQL database is a good way to go. It allows you to use common dplyr functions such as select(), filter(), mutate(), group_by(), summarise() and arrange() with a SQL database.
+
+As with sending a SQL script, you'll need to create the connection to the database first:
+
+```
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:328]: Done

File: learning-development/r.qmd [hunk 8]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+                      Database = "database_name",
+                      UID = "",
+                      PWD = "",
+                      Trusted_Connection = "Yes"
+)
+```
+
+Once you've established a connection, you can then create piped queries as you would with dplyr:
+
+```
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:348]: Done

File: learning-development/r.qmd [hunk 9]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+- we use tbl() to point to the SQL table on the SQL database in place of supplying a data-frame
+- we use the collect() function to tell R to collate the commands and run them on the SQL database
+
+### Common issues interacting with a SQL database from within R
+
+#### DBI varchar(max) column ordering bug
+
+A known bug in DBI is that if a SQL table contains columns of type varchar(max), then any queries including those columns will fail if those columns aren't the last to be called in the query. This is a bit esoteric, but you can identify this issue if you're getting the following error message when you run your query:
+
+```
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:367]: Done

File: learning-development/r.qmd [hunk 10]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+In dbClearResult(rs) : Result already cleared
+```
+
+There are a couple of solutions to this:
+- avoid creating tables containing varchar(max) columns (i.e. always set a field length for your character fields)
+- make sure all varchar(max) columns are placed last in ay select query that you write
+
+This latter route can be automated using the following function: 
+
+```
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:379]: Done

File: learning-development/r.qmd [hunk 11]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+SET ANSI_PADDING OFF
+SET NOCOUNT ON;
+```
+
+---
+
+### Using dbplyr to create database queries
+
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:323]: Done

File: learning-development/r.qmd [hunk 12]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+---
+
+## Interacting with a SQL database from within R-scripts
 
 ---
 
+R can be used to execute SQL scripts to extract data from a database as well as querying the database directly via R. There are two primary ways to do this: 1. executing a SQL script as is within R and 2. using dbplyr to create SQL code for you. Which you use will depend on how comfortable with SQL and R and also if you already have existing SQL scripts that you want to execute or you're writing new database queries.
+
+### Executing SQL scripts from within R
+
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:260]: Done

File: learning-development/r.qmd [hunk 13]
```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
...
+```
+SET ANSI_PADDING OFF
+SET NOCOUNT ON;
+```
+
+---
+
+### Using dbplyr to create database queries
+
+If you're familiar with dplyr then using the dbplyr package to interact with a SQL database is a good way to go. It allows you to use common dplyr functions such as select(), filter(), mutate(), group_by(), summarise() and arrange() with a SQL database.
```
[the remark this answers is not in the document]
Comment (rmbielby) [reply learning-development/r.qmd:324]: Done

File: learning-development/r.qmd (hunk 1 above)
Comment (rmbielby) [reply learning-development/r.qmd:353]: Both syntax styles work identically. Either way, it collates the commands and then collects the data. Each of the different styles probably make different sense to different people. I just put everything inside the collect function as I thought that fitted better conceptually with the idea that it's evaluating all the rest of the commands that preceded it in one go, whereas the pipe notation kind of makes it look like each command is fully run sequentially.

File: learning-development/r.qmd (hunk 2 above)
Comment (rmbielby) [reply learning-development/r.qmd:361]: Done

File: learning-development/r.qmd (hunk 3 above)
Comment (rmbielby) [reply learning-development/r.qmd:375]: Done

**Review by cjrace — APPROVED** [review]

**Pull request merged**